### ISSUE TITLE
Consider making the build script version check optional

### ISSUE THREAD

**Issue by shepmaster** [core-error/core-error#8]
As this crate intends to be a foundational one, having the ability to have minimal dependencies is a good quality. Additionally, I'm not a *fan* of automatically detecting the Rust version as I believe it leads to sloppier downstream code (a variant of "it worked on my computer", effectively).

I'd appreciate it if the usage of `version_check` was made optional and the corresponding version flags were accessible.

**Comment by roblabla** [comment]
This is loosely related to #1.

> having the ability to have minimal dependencies is a good quality.

I agree, and I chose `version_check` to do the rust version checking explicitly because it has no dependencies. This crate will never have any additional dependencies - and I might do away with `version_check` if it ever becomes a bottleneck in compile times.

> I'm not a fan of automatically detecting the Rust version as I believe it leads to sloppier downstream code (a variant of "it worked on my computer", effectively).

Can you elaborate on this a bit? I don't really see how asking the user to specify the rust version through a feature would lead to less sloppy code.

**Comment by shepmaster** [comment]
> Can you elaborate on this a bit?

I view it as a facet similar to [my comment about minimal version testing](https://github.com/core-error/core-error/issues/2#issuecomment-539679394) as well as the general idea of testing with the earliest supported Rust version. If I happen to develop with Rust 1.50 which magically enables feature Y in a crate, then I don't realize that it doesn't exist in Rust 1.40 until too late (when it gets to CI, ideally, or to an end-user, less ideally).

If, on the other hand, I go to use a feature Y from the crate and it's missing (or better yet documented with what version / feature flag I need to set), then I can make an informed decision to spend the time/energy about further development of my feature that would have used feature Y.

**Comment by roblabla** [comment]
I understand, but in this case, all we’re doing is adding impls on a type that exists anyways, since it’s in the libcore. No functionality is being added depending on the rust version, only an error impl. So the code was already targeting the wrong version before core-error got involved - and it’s not our job to tell the user that they’re using the wrong rust APIs.

IMO it would be much more confusing to end up in a situation where a certain libcore error does not implement Error because of a missing feature flag. The type would exist but not implement Error, but only in nostd build (since std builds just use the trait from the libstd). That’s a special kind of debugging nightmare.

As a general rule, crates targeting rust 1.40 should have a `rust-toolchain` file next to the Cargo.toml with the string “1.40.0”. Then developers will never mess up, on-boarding of new contributors is made simple, and it also makes a nice place to document the MSRV and track its evolution in the git history.

---

There are currently two functionalities gated on a rust version besides the error impls, TypeInfo::type_name gated on rust 1.38 and downcasting gated on Rust 1.7. I’ll probably remove the former as we don’t use it and it’s otherwise doc(hidden), while the later is gated on such an old version that I feel it probably doesn’t matter beyond a note in the docs.

I wish downcasting worked on 1.6 (the earliest nostd version) but that release had a botched Any impl, so amen to that.

EDIT: TypeInfo::type_name is removed now.

**Comment by roblabla** [comment]
So I'm looking at core-error again. The idea of making the rustc version configurable through cargo features is a definite no-go.

1. features are settable by libraries, and I can guarantee someone somewhere will set them in a library and cause pain and frustration
2. Setting the version too high will cause build errors. Setting the version too low will cause some errors to mysteriously not impl the error trait, causing confusion.

Taking a step back, the only valid reason I see for removing the build script is build performance. Auto-setting won't lead to sloppy code because we're not adding types, only impls. Unfortunately, in the absence of cfg(version), I don't see a way of getting rid of the build script, so it will have to stay for now.

Closing as wontfix.
